src/main/kotlin/org/rust/lang/core/thir/ThirExpr.kt

```diff
@@ -0,0 +1,122 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.thir
+
+import org.rust.lang.core.mir.schemas.MirBorrowKind
+import org.rust.lang.core.mir.schemas.MirSpan
+import org.rust.lang.core.psi.RsLitExpr
+import org.rust.lang.core.psi.RsStructItem
+import org.rust.lang.core.psi.ext.ArithmeticOp
+import org.rust.lang.core.psi.ext.LogicOp
+import org.rust.lang.core.psi.ext.UnaryOperator
+import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.regions.Scope as RegionScope
+
+sealed class ThirExpr(val ty: Ty, val span: MirSpan) {
+    class Scope(
+        val regionScope: RegionScope,
+        val expr: ThirExpr,
+        ty: Ty,
+        span: MirSpan,
+    ) : ThirExpr(ty, span)
```

Comment:
From my experience, it usually more clear to override abstract vals then forward parameters to a super constructor

```suggestion
sealed class ThirExpr() {
    abstract val ty: Ty
    abstract val span: MirSpan

    class Scope(
        val regionScope: RegionScope,
        val expr: ThirExpr,
        override val ty: Ty,
        override val span: MirSpan,
    ) : ThirExpr()
```